distributed trainning for transformer

xuezhong opened the pull request:
distributed trainning for transformer,two options:
1 device cpu, which can run only with cpu instead of GPU
2 local, which can run on multi machines

jacquesqiao committed:
distributed trainning for transformer

panyx0718 commented on fluid/neural_machine_translation/transformer/train.py:
```diff
@@ -257,167 +246,267 @@ def train(args):
+        optimizer.minimize(sum_cost)
+
+    def train_loop(exe, train_progm):
+        def read_multiple(reader,
```
It seems there is no need to move read_multiple and split_data to here?

xuezhong replied:
I move these functions out

panyx0718 commented on fluid/neural_machine_translation/transformer/train.py:
```diff
@@ -257,167 +246,267 @@ def train(args):
+        # Initialize the parameters.
+        if TrainTaskConfig.ckpt_path:
+            fluid.io.load_persistables(exe, TrainTaskConfig.ckpt_path)
+            #lr_scheduler.current_steps = TrainTaskConfig.start_step
```
why is this line commented?

xuezhong replied:
a mistake, fix it

panyx0718 commented on fluid/neural_machine_translation/transformer/train.py:
```diff
@@ -257,167 +246,267 @@ def train(args):
+            epsilon=TrainTaskConfig.eps)
+        optimizer.minimize(sum_cost)
+
+    def train_loop(exe, train_progm):
```
this function seems too long. Maybe it should be moved out of of train?

xuezhong replied:
ok

panyx0718 reviewed (commented):
Can you simply ```train``` function? For example, there is no need to put ```read_multiple``` and ```train_loop``` inside ```train```? ```train``` is not readable now.

jacquesqiao committed:
add sync mode option

jacquesqiao committed:
merge
Merge branch 'develop' of https://github.com/PaddlePaddle/models into transformer_merge

jacquesqiao committed:
change code arrangment

Commit:
add readme doc for distributed trainning

kuke commented on fluid/neural_machine_translation/transformer/README_cn.md:
````diff
@@ -154,9 +154,64 @@ perl multi-bleu.perl data/newstest2013.tok.de < predict.tok.txt
 ```
 BLEU = 25.08, 58.3/31.5/19.6/12.6 (BP=0.966, ratio=0.967, hyp_len=61321, ref_len=63412)
 ```
+###分布式训练
````
<img width="927" alt="2018-07-10 3 57 16" src="https://user-images.githubusercontent.com/3064195/42496899-109b9fc8-845a-11e8-93e0-0fef6b4bd508.png">

Please  pay attention to the format

xuezhong replied:
ok

Commit:
update readme

kuke commented on fluid/neural_machine_translation/transformer/README_cn.md:
```diff
@@ -154,9 +154,64 @@ perl multi-bleu.perl data/newstest2013.tok.de < predict.tok.txt
+###分布式训练
+transformer 模型支持同步或者异步的分布式训练。分布式的配置主要两个方面:
+1 命令行配置
+  --local参数，这个参数有两个取值，True表示单机训练，而False表示使用分布式训练。默认为单机训练模式。
```
Please use `` for English names and values in text.

xuezhong replied:
ok

panyx0718 commented on fluid/neural_machine_translation/transformer/train.py:
```diff
@@ -398,6 +404,7 @@ def test(exe=test_exe):
                 feed_dict[sum_cost.name + "@GRAD"] = 1. / total_num_token
             outs = train_exe.run(fetch_list=[sum_cost.name, token_num.name],
                                  feed=feed_list)
+            train_exe.bcast_params()
```
这个不需要了吧？

xuezhong replied:
en  remove it

Commit:
update readme doc

Commit:
merge

Commit:
update readme doc

Commit:
need nott bcast_param

panyx0718 commented:
@xuezhong 找 @kuke 确认一下吧

kuke commented:
LGTM
